Framework/src/RRNG.cc

```diff
@@ -0,0 +1,131 @@
+#include "PandaTree/Framework/interface/RRNG.h"
+#include "Math/QuantFuncMathCore.h"
+#include "TString.h"
+#include <iostream>
+#include <stdexcept>
+#include <cassert>
+
+using namespace std;
+using namespace panda;
+
+int RRNG::_default_idx;
+
+RRNG::RRNG(int seed, int maxEntropy): 
+  _rng(seed),
+  _maxEntropy(maxEntropy)
+{
+  assert(_maxEntropy > 0);
+  _default_idx = -1;
+  _x = new double[_maxEntropy];
+  _available = new bool[_maxEntropy];
+  generate();
+}
+
+RRNG::RRNG(RRNG const& other): 
+  _rng(other._rng),
+  _maxEntropy(other._maxEntropy)
+{
+  _x = new double[_maxEntropy];
+  _available = new bool[_maxEntropy];
+  generate();
+}
+
+void RRNG::generate() 
+{
+  _rng.RndmArray(_maxEntropy, _x);
+  for (int i = 0; i != _maxEntropy; ++i) 
+    _available[i] = true;
+}
+
+int RRNG::_nextAvailable(int start)
+{
+  for (int i = start; i != _maxEntropy; ++i) {
+    if (_available[i])
+      return i;
+  }
+  return -1;
+}
+
+double RRNG::_requestNumber(int& idx) 
+{
+  int idx_;
```

Comment:
This is where we should actually check for 0 size of pre-generated numbers. If size is 0, just return a new number from the `TRandom3` instance.